[#10920] Add front-end tests for instructor courses page

halfwhole opened the pull request:
Part of #10920

**Outline of Solution**
- Added API test for loading courses (active, archived, and soft deleted)
- Added API test for getting course statistics
- Added tests for disabling front-end buttons when instructor has no modification permissions
- Added tests for sorting by column

t-cheepeng reviewed (commented):
Great Work!

Let's add some more tests
- Snapshot tests for the boolean variables which affect the layout of the page
- Unit tests to ensure that service methods are called properly with proper params

halfwhole commented:
Added more tests here:
- Added unit tests for archiving/unarchiving/soft deleting/permanent deleting
- Added testing for service methods being called with correct params
- Added snapshot tests for two more boolean variables (`isAddNewCourseFormExpanded` and `isArchivedCourseExpanded`), other boolean variables should already be covered by previous snapshot tests

t-cheepeng reviewed (approved):
LGTM

ChooJeremy commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -292,10 +292,10 @@ export class InstructorCoursesPageComponent implements OnInit {
    * Permanently deletes a soft-deleted course in Recycle Bin.
    */
-  onDeletePermanently(courseId: string): void {
+  onDeletePermanently(courseId: string): Promise<void> {
```
What's the goal of returning Promise here?

halfwhole replied:
The promise helps us to know if the function has finished deleting the course (because the ngbmodal result itself is a promise) -- otherwise, it would be hard for the tests to know when the async operations are done before testing for the desired changes. There doesn't seem to be any other way to do it that's as straightforward as using promises, so I went with this approach

ChooJeremy reviewed (approved):
LGTM
